### DISCUSSION TITLE
DirectPlay IPX over LAN using NE2000

### DISCUSSION THREAD

**Discussion by Tablaski** [joncampbell123/dosbox-x#5289] [General]
I did read @joncampbell123 in https://github.com/joncampbell123/dosbox-x/issues/4263 but I don't understand how he has done to link two DOSBOX-X instances using NE2000 emulation to play multiplayer together through the IPX tunneling

I'm trying to run Z for windows 95 which I've installed DirectX 7.0 on, to play through IPX. The game uses DirectPlay.
Both PCs are set up with NE2000 emulation and pcap back-end. We tried a couple of real adapters settings 
(Ethernet, Wi-FI, loopback adapters) to no avail.

The IPX tunneling seems to work over our internet box LAN. Before booting on Windows 95, using "ipxnet status" I got : 
- Connected status on client side 
- Server Active status on server side + the IP of the client in the list of IPs.

IPX/SPX protocol is installed on w95 (from the vanilla drivers of the O/S, not drivers from an Internet source). 
Tried playing with different settings (unique network adress, Frame type Ethernet 802.2 notably), nothing.

Read about IPXwrapper, enet encapsulation (https://github.com/dosbox-staging/dosbox-staging/issues/947 which concludes that IPX tunneling works anyway)... not sure what to do... 

We tried the connexion using the 01.03.2024 version of Dosbox-X though, I don't know if the IPXtunnel evolved on newer versions ?
NB : I didn't try port forwarding on the internet box but I've check TCP/UDP ports were open on each PC's firewalls

Any help please ?

**Comment by joncampbell123** [comment]
You need to enable both NE2000 and IPX emulation in dosbox.conf before booting Windows 95.
Then there is the dosbox.conf option to tell NE2000 emulation to redirect all IPX traffic to/from the IPX tunnel.
It should still work, I haven't changed anything in that source code since finishing that feature.
DOSBox-X extends the IPX tunnel protocol slightly to allow this trick to work so don't expect it to work with other DOSBox forks.

**Comment by Tablaski** [comment]
Thanks for your reply
I'm not sure what you mean by "the dosbox.conf option to tell NE2000 emulation to redirect all IPX traffic". 
Do you mean network redirector = true ?

Is the ne2000 macaddr value important to set other than random ? Should it match the actual "realnic" adapter ?

**There is these default settings in the dosbox-x.conf which is launched initially :**

network redirector = true

[ne2000]
ne2000  = false
nicbase = 300
nicirq  = 3
macaddr = random
backend = auto

**And then I restart dosbox using a win95.conf file with these settings before booting :** 

[ne2000]
ne2000=true
nicirq=10
backend=pcap

[ethernet, pcap]
realnic=1

[ipx]
ipx=true

[autoexec]
ipxnet startserver 19990 (OR ipxnet connect 192.168.1.xx 19990 depending on server or client)

(and then I do boot c: after checking the ipxnet status)

**Comment by joncampbell123** [comment]
Huh, I thought I had added an option to control whether NE2000 redirects IPX packets to the tunnel.

**Comment by joncampbell123** [comment]
Yes, MAC address should be random. That was added because there used to be a default MAC address but that caused collisions that prevented communication between multiple versions of DOSBox-X unless you went out of your way to change it.

**Comment by joncampbell123** [comment]
I also want to point out that the IPX-NE2000 tunnel code so far was only tested with two programs.

The "WINPOPUP" program in Windows 95/98, and

Command & Conquer multiplayer (Windows 95 to Windows 95, or Windows 95 to MS-DOS).

It should work with anything else that uses IPX in a guest environment, but that hasn't been fully tested.

**Comment by Tablaski** [comment]
I would be happy to test it with Z95 (the Z game for windows 95) then, if it's possible :-)

**Comment by Tablaski** [comment]
@joncampbell123 I've managed to find and enable your ne2k_ipx_redirect setting , toggle many LOG_MSG and recompile using Visual Studio

I seem to get many packets at startup, some read some sent (might help that I enabled netbios ?). Also when I use WinPopup (haven't been able to send through the IPX Tunnel though). I even got some "sent to IPX" at some point so the loggers are working.

But when I use Directplay IPX (Directx 7), whether from the game itself or testing from dxdiag.exe, nothing happens. No packet logged at all. Even if not recognized probably as IPX my understanding is I would set a log message saying they were sent to Ethernet. Tried using win 95 and win 98

Any idea ? What DirectX/Directplay did you use when testing with C&C 95 ? Where you using PCAP as well ?

**Reply by Torinde** [reply]
Great. Can you please make a PR, so that this option is available as user .conf setting without recompilation?

**Reply by Tablaski** [reply]
I'm sorry but I'm absolutely not a developer, even less a C++ one. 
That was an easy fix here but I don't really know how to tweak options.
I could perhaps figure out but that would be way faster for someone else, and I plan rather on packaging the recompiled DosBox-X + Z95 preinstalled on windows 98 and ready to play :-)

**Comment by Tablaski** [comment]
I reinstalled the game with its original DirectX 5 and now when the client game tries to join the server it does send constantly packets that are forwarded to IPX. 

I feel i'm getting very close now but I still don't see the host game from the client. Both NE2000 when set to work with the WiFI 6 real adapter receive constant packets which might mean they indeed communicate with the network ?

I made sure the ipx tunnel connected on both side before launching both windows 95 but then I was testing with two instances on the same PC

**Comment by Tablaski** [comment]
I MADE IT !!! i didn't know i had the same mac address twice ! you can add Z 95 to C&C 95 now @joncampbell123

**Reply by joncampbell123** [reply]
> I MADE IT !!! i didn't know i had the same mac address twice ! you can add Z 95 to C&C 95 now @joncampbell123

Ethernet cards have the same conflict problem that happens if two machines on a LAN have the same IP address. They have to be different or they can't talk.

**Reply by Tablaski** [reply]
Sure. The thing is I had this in the default conf :

ne2000  = false
nicbase = 300
nicirq  = 3
macaddr = random
backend = auto

And then I reload my own conf to mount images, etc, including :

[ne2000]
ne2000=true
nicirq=10
backend=pcap

And so it seems, doing that, the macaddr = random setting was lost

That's why it took me a while to notice, because I was working under win95 where there is no proper networking tools.
When I switched to win98, ipconfig /all shown me this problem

Thanks for your support, that NE2K <-> IPXnet redirection is great, please include it asap in the config :-)

**Comment by joncampbell123** [comment]
I'm considering another addition to take how IPX tunneling works and implement one for NE2000 Ethernet so multiple instances can talk to each other. The tunnel would carry raw Ethernet packets instead of IPX packets.

**Reply by Tablaski** [reply]
Nice, we could network-play anything then !

**Reply by simonchen** [reply]
Npcap already did transfer the raw Ethernet packets (frames) over physical network interfaces like you realized backend=pcap
you don't have to spend time to do the same copy job on Npcap...
Even you do the copy job, you can't guarantee that IPX packets couldn't be thrown alway from any physical network interfaces since the L2 IPX has been obsoleted for many years.

**Comment by Torinde** [comment]
Related:
- https://github.com/dosbox-staging/dosbox-staging/issues/947
